optaplanner-core/src/main/java/org/optaplanner/core/impl/util/ConsecutiveIntervalData.java

```diff
@@ -0,0 +1,107 @@
+        if (!intersectedIntervalClusterMap.isEmpty()
+                && intersectedIntervalClusterMap.get(intersectedIntervalClusterMap.firstKey()).getEndSplitPoint()
+                        .compareTo(interval.getStartSplitPoint()) < 0) {
+            intersectedIntervalClusterMap = intersectedIntervalClusterMap.subMap(intersectedIntervalClusterMap.firstKey(),
+                    false, intersectedIntervalClusterMap.lastKey(), true);
+        }
+
+        if (intersectedIntervalClusterMap.isEmpty()) {
+            IntervalSplitPoint<_IntervalValue, _PointValue> start = splitPointSet.floor(interval.getStartSplitPoint());
+            clusterStartSplitPointToCluster.put(start, new IntervalCluster<>(splitPointSet, start));
+            return;
+        }
+        IntervalCluster<_IntervalValue, _PointValue> intervalCluster =
+                intersectedIntervalClusterMap.get(intersectedIntervalClusterMap.firstKey());
+        IntervalSplitPoint<_IntervalValue, _PointValue> oldStart = intervalCluster.getStartSplitPoint();
+        intervalCluster.addInterval(interval);
+        intersectedIntervalClusterMap.tailMap(intersectedIntervalClusterMap.firstKey(), false).values()
+                .forEach(intervalCluster::mergeIntervalCluster);
+        intersectedIntervalClusterMap.tailMap(intersectedIntervalClusterMap.firstKey(), false).clear();
+        if (oldStart.compareTo(intervalCluster.getStartSplitPoint()) > 0) {
+            clusterStartSplitPointToCluster.remove(oldStart);
+            clusterStartSplitPointToCluster.put(intervalCluster.getStartSplitPoint(), intervalCluster);
+        }
+    }
+
+    protected void removalInterval(Interval<_IntervalValue, _PointValue> interval) {
+        Map.Entry<IntervalSplitPoint<_IntervalValue, _PointValue>, IntervalCluster<_IntervalValue, _PointValue>> intervalClusterEntry =
+                clusterStartSplitPointToCluster.floorEntry(interval.getStartSplitPoint());
+        IntervalCluster<_IntervalValue, _PointValue> intervalCluster = intervalClusterEntry.getValue();
+        clusterStartSplitPointToCluster.remove(intervalClusterEntry.getKey());
+        for (IntervalCluster<_IntervalValue, _PointValue> newIntervalCluster : intervalCluster.removeInterval(interval)) {
+            clusterStartSplitPointToCluster.put(newIntervalCluster.getStartSplitPoint(), newIntervalCluster);
+        }
+    }
+
+    public List<IntervalCluster<_IntervalValue, _PointValue>> getIntervalClusters() {
+        return valueList;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
+        ConsecutiveIntervalData<?, ?> that = (ConsecutiveIntervalData<?, ?>) o;
+        return valueList.equals(that.valueList);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(valueList);
+    }
```

Comment:
Same question on what we need equality for here.

Reply:
Same as above; tests